### PR TITLE
[EMP-779] Volume migration to io2 script for EMP

### PR DESCRIPTION
**Pull request by NeelavaChatterjee**
## Issue(s)
[EMP-779](https://platform9.atlassian.net/browse/EMP-779): prepare basic migration script for migration to io2 volumes

## Summary
Volume migration script for EMP which can be used to migrate volume of any type to io2. 

## Usage
`python3 migrate_to_io2.py --aws_access_key_id AWS_ACCESS_KEY_ID --aws_secret_access_key AWS_SECRET_ACCESS_KEY --region_id AWS_REGION_ID --volume_ids vol-1xxxxxxxxxxxxxx vol-2xxxxxxxxxxxxxx`
## Testing
Manually migrated volumes to io2 using the script and checked AWS console for confirmation.
```bash
ubuntu@neelava-dev-vm:~/support-locker/emp$ python3 migrate_to_io2.py --aws_access_key_id AWS_ACCESS_KEY_ID --aws_secret_access_key AWS_SECRET_ACCESS_KEY --region_id AWS_REGION_ID --volume_ids vol-xxxxxxxxxxxxxxxx
[2023-09-25 06:40:24,303] INFO - Modifying volume vol-xxxxxxxxxxxxxxxx to io2
[2023-09-25 06:40:24,548] INFO - Modifying volume vol-xxxxxxxxxxxxxxxx to io2 with the following parameters: iops: 40000
[2023-09-25 06:40:24,694] INFO - Fetching modfication status for volume [vol-xxxxxxxxxxxxxxxx].
[2023-09-25 06:40:54,789] INFO - Polling volume(vol-xxxxxxxxxxxxxxxx) status: completed, progress: 100
[2023-09-25 06:40:54,789] INFO - Volume modified: vol-xxxxxxxxxxxxxxxx: completed. Total timetaken: 0:00:03
[2023-09-25 06:40:54,790] INFO - Successfully modified volume vol-xxxxxxxxxxxxxxxx to io2
```

[EMP-779]: https://platform9.atlassian.net/browse/EMP-779?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: emp/migrate_to_io2.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            t.join()
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Modifies an EBS volume to io2 and sets the IOPS')
+
+    parser.add_argument('--aws_access_key_id', required=True, help='The AWS access key ID.')
+    parser.add_argument('--aws_secret_access_key', required=True, help='The AWS secret access key.')
+    parser.add_argument('--region_id', required=True, help='The AWS region id.')
+    parser.add_argument('--iops', type=int, help='The IOPS to set for the volume.')
+    parser.add_argument('--volume_ids', required=True, nargs='+', help='The IDs of the EBS volumes to modify.')
```
Comment (OmkarDeshpande7) [inline emp/migrate_to_io2.py:91]: fail if any of these values is not provided

Comment (OmkarDeshpande7) [reply emp/migrate_to_io2.py:91]: or are empty

File: emp/migrate_to_io2.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                VolumeId=volume_id,
+                VolumeType='io2',
+                Iops=iops
+            )
+
+            self.check_modification_status(volume_id)
+
+            logger.info('Successfully modified volume %s to io2', volume_id)
+        except Exception as e:
+            logger.error('Failed to modify volume %s to io2: %s', volume_id, e)
```
Comment (OmkarDeshpande7) [inline emp/migrate_to_io2.py:69]: please see if you can have granular exception handling

File: emp/migrate_to_io2.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+)
+logger = logging.getLogger('PF9')
+
+class MigrateVolume:
+    def __init__(self):
+        self.ec2_client = boto3.client('ec2', aws_access_key_id=AWS_ACCESS_KEY_ID, aws_secret_access_key=AWS_SECRET_ACCESS_KEY, region_name=AWS_REGION_ID)
+        self.polltime = 30
+
+    def calculate_max_iops(self, volume_size):
+        max_iops = min(math.ceil(volume_size * 500), 64000)
```
Comment (OmkarDeshpande7) [inline emp/migrate_to_io2.py:25]: extract 500 to default iops per GB. and get that as flag (not required value)

File: emp/migrate_to_io2.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        logger.info("Fetching modfication status for volume [%s].", volume_id)
+        try:
+            response = self.ec2_client.describe_volumes_modifications(VolumeIds=[volume_id])
+        except botocore.exceptions.ClientError as err:
+            if err.response["Error"]["Code"] == "InvalidVolumeModification.NotFound":
+                logger.warning(err.response["Error"]["Message"])
+                return
+            raise err
+
+        while response["VolumesModifications"][0]["ModificationState"] != "completed" and response["VolumesModifications"][0]["ModificationState"] != "failed":
```
Comment (OmkarDeshpande7) [inline emp/migrate_to_io2.py:38]: check possible NPE? `response["VolumesModifications"][0]`

Comment (NeelavaChatterjee) [reply emp/migrate_to_io2.py:42]: Done.

File: emp/migrate_to_io2.py (hunk 3 above)
Comment (NeelavaChatterjee) [reply emp/migrate_to_io2.py]: Done

File: emp/migrate_to_io2.py (hunk 1 above)
Comment (NeelavaChatterjee) [reply emp/migrate_to_io2.py]: Done

File: emp/migrate_to_io2.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            # Calculate iops based on the volume size and IOPS/GB
+            iops = self.calculate_iops(volume_info['Size'], iops_per_gb)
+
+            logger.info('Modifying volume %s to io2 with the following parameters:- IOPS/GB: %s, IOPS: %s', volume_id, iops_per_gb, iops)
+            # Modify the volume to io2 and set the IOPS and throughput.
+            self.ec2_client.modify_volume(
+                VolumeId=volume_id,
+                VolumeType='io2',
+                Iops=iops
+            )
```
Comment (OmkarDeshpande7) [inline emp/migrate_to_io2.py:71]: Add `MultiAttachEnabled=True` here. We want to modify volume to io2 and make them multi-attach enabled.

**Review by OmkarDeshpande7 — COMMENTED** [review]
looks good. Just one comment

File: emp/migrate_to_io2.py (hunk 5 above)
Comment (NeelavaChatterjee) [reply emp/migrate_to_io2.py]: done

**Review by OmkarDeshpande7 — APPROVED** [review]

**Review by atiwaripw — APPROVED** [review]

**Pull request merged**